## Fechas y horas

El módulo **datetime** integrado de Python ofrece los tipos datetime, date y time. El tipo datetime combina la información almacenada en date y time y es el más utilizado:


In [2]:
from datetime import datetime, date, time

In [3]:
dt = datetime(2011, 10, 29, 20, 30, 21)

In [4]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

In [5]:
dt.day # dt.minute

29

In [10]:
type(dt.second)

int

In [11]:
dt.date() #dt.time()

datetime.date(2011, 10, 29)

In [12]:
dt.time()

datetime.time(20, 30, 21)

El método **strftime** formatea un datetime como cadena de texto:

In [13]:
dt.strftime('%Y-%m-%d %H:%M')

'2011-10-29 20:30'

In [ ]:
a=

Las cadenas de texto se pueden convertir en objetos datetime con la función **strptime**:

In [14]:
datetime.strptime('20091031', '%Y%m%d')

datetime.datetime(2009, 10, 31, 0, 0)

Cuando estamos agregando o agrupando de otro modo datos de series temporales, de vez en cuando es útil reemplazar campos de hora de una serie de datetimes (por ejemplo, sustituyendo los campos minute y second por cero):

In [17]:
dt_hour = dt.replace(minute=0, second=1) 

In [18]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

In [19]:
dt_hour

datetime.datetime(2011, 10, 29, 20, 0, 1)

In [9]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

In [20]:
dt_hour = dt.replace(hour=11,minute=4, second=3)

In [21]:
dt_hour

datetime.datetime(2011, 10, 29, 11, 4, 3)

La diferencia de dos objetos datetime produce un tipo **datetime.timedelta**:

In [23]:
dt2 = datetime(2011, 11, 15, 22, 30)
dt2

datetime.datetime(2011, 11, 15, 22, 30)

In [24]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

In [25]:
delta = dt2-dt

delta

datetime.timedelta(days=17, seconds=7179)

El resultado timedelta(17, 7179) indica que el timedelta codifica un desplazamiento de 17 días y 7,179 segundos.

Sumar un timedelta a un datetime produce un nuevo datetime movido de su sitio:

In [26]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

In [27]:
dt+delta

datetime.datetime(2011, 11, 15, 22, 30)

¿Cuantos días has vivido?

In [32]:
f=datetime(2040, 4, 5, 22, 30)
na=datetime(1988, 11, 5, 22, 30)

In [33]:
d=f-na

In [34]:
d

datetime.timedelta(days=18779)

In [36]:
d.days/365

51.44931506849315

# Control de Flujo

Python tiene varias palabras clave internas para lógica condicional, bucles y otros conceptos estándares de control de flujo, que pueden encontrarse en otros lenguajes de programación.

## if, elif y else

La sentencia **if** es uno de los tipos de sentencia de control de flujo más conocidos. Comprueba una condición que, si es *True*, evalúa el código del bloque que le sigue:


In [38]:
x = -5
if x < 0:
    print('x es negativo')

x es negativo


Una sentencia *if* puede ir seguida de manera opcional por uno o más bloques **elif** y un bloque multifuncional **else** si todas las condiciones son False:

In [40]:
x = 20

if x < 0:
    print('x es negativo')
elif x==0:
    print('x es igual a cero')
elif 0 < x < 15:
    print('x es positivo pero menor que 15')
else:
    print('x es positivo mayor o igual a 15')

x es positivo mayor o igual a 15


 Con una condición compuesta utilizando **and** o **or**, las condiciones se evalúan de izquierda a derecha 

In [42]:
a = 5; b = 7

c = 8; d = 4

In [45]:
if a < b and d > c:
    print('Cierto')

# Bucles for

Los bucles **for** se utilizan para aplicar determinadas instrucciones a una colección (como una lista o una tupla) o a un iterador.

In [46]:
L = [1, 2, None, 4, None, 5]

for val in L:
    print(val)

1
2
None
4
None
5


Se puede avanzar un bucle for a la siguiente repetición, saltando el resto del bloque, mediante la palabra clave **continue**.

In [48]:
a=3

In [49]:
a+=4

In [50]:
a

7

In [51]:
L = [1, 2, None, 4, None, 5]
total = 0

for val in L:
    total += val

TypeError: unsupported operand type(s) for +=: 'int' and 'NoneType'

In [55]:
L = [1, 2, None, 4, None, 5]
total = 0

for val in L:
    if val is None:continue
    total += val

In [53]:
total

12

Un bucle for puede darse por terminado también con la palabra clave **break**.

In [56]:
seq = [1, 2, 0, 4, 6, 5, 2, 1]
    
total_antes_5 = 0

for val in seq:
    if val == 5: break
    total_antes_5 += val



In [57]:
total_antes_5

13

# Bucles while

Un bucle while especifica una condición y un bloque de código que se va ejecutar hasta que la condición evalúe a **False** o el bucle sea finalizado explícitamente con **break**:


In [58]:
x = 50
total = 0

while x > 0:
    if total > 500: break
    print(total)
    total += x
    

0
50
100
150
200
250
300
350
400
450
500


**pass** es la sentencia «no operativa» (es decir, que no hace nada) de Python. Se puede utilizar en los bloques en los que no hay que realizar ninguna acción

In [61]:
x=0

if x < 0:
    print("negative")
elif x == 0:
    pass
else:
    print("positive!")

La función **range** genera una secuencia de enteros espaciados por igual:

In [62]:
range(11)

range(0, 11)

In [64]:
list(range(11))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Se puede dar un inicio, un final y un paso o incremento (que puede ser negativo):

In [30]:
list(range(0, 20, 2))

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

In [67]:
list(range(12, 0, -2))

[12, 10, 8, 6, 4, 2]

**range** produce enteros hasta el valor final pero sin incluirlo; se suele utilizar para aplicar instrucciones a distintas secuencias según un índice:


In [69]:
seq = [1, 4, 7, 23]

for i in range(len(seq)):
    print(f'elemento {i}: {seq[i]}')

elemento 0: 1
elemento 1: 4
elemento 2: 7
elemento 3: 23


In [71]:
range(len(seq))

range(0, 4)

In [72]:
sum(seq)

35

In [73]:
(100*101)/2

5050.0

In [76]:
sum(range(0,100,2))




2450

In [75]:
range(0,101)

range(0, 101)

 suma de los cuadrados de enteros


In [77]:
P=[]

for i in range(101):
    P.append(i**2)
    

In [79]:
sum(P)

338350